Add Docker benchmark suite

sandrokeil opened the pull request:
Please merge #1 first.

I guess the ArangoDB benchmark is not working properly, because I don't get the output of the writes. I get only:

```
Writer 140136754645736 started
Projection 140136754551528 started
``` 

@prolic Any ideas, because the other tests are working and there are collections. But there are no data for the real world test in the database. Looks like an exception in the threads?

prolic commented:
I will need to check why your output is not showing correctly. Two ideas that I have right now:
1) Do you run the benchmark on PHP 7.2?
2) Do you use pthreads self-compiled from latest master?
Both is required in order to run for the benchmarks. Multithreading support is broken in PHP 7.0 / 7.1 and pthreads for 7.2 is not yet finished, but current master works good enough for our benchmarks.

sandrokeil commented:
@prolic Yep, that's the precondition. We have a new [prooph Docker Image](https://github.com/prooph/docker-files/blob/master/php/7.2-zts#L57-L63) for the benchmarks.

prolic commented:
ok, will have to check later in more detail

prolic commented:
are the other benchmarks running for you? so only arangodb is a problem right now?

sandrokeil commented:
@prolic I guess it's a ArangoDB Client connection problem, I try to figure out why it's break.

sandrokeil commented:
That's the error, I don't know why this happend.

```
Undefined class constant 'DEFAULT_CIPHERS'
#0 /app/vendor/triagens/arangodb/lib/ArangoDBClient/ConnectionOptions.php(205): ArangoDBClient\ConnectionOptions::getDefaults()
#1 /app/vendor/triagens/arangodb/lib/ArangoDBClient/Connection.php(112): ArangoDBClient\ConnectionOptions->__construct(Array)
#2 /app/src/functions.php(86): ArangoDBClient\Connection->__construct(Array)
#3 /app/src/StreamCreator.php(37): Prooph\EventStoreBenchmarks\createConnection('arangodb')
#4 [internal function]: Prooph\EventStoreBenchmarks\StreamCreator->run()
```

sandrokeil commented:
@prolic I've solved the issues

prolic commented on src/benchmark.php:
```diff
@@ -35,6 +35,16 @@
+
+    $filenameArangoDb = '../vendor/triagens/arangodb/lib/ArangoDBClient/ConnectionOptions.php';
+    $contentArangoDb = file_get_contents($filenameArangoDb);
+    $newContentArangoDb = str_replace('DefaultValues::DEFAULT_CIPHERS', 'null', $contentArangoDb);
```
I think that's not needed, you have to include that class `DefaultValues` manually before a thread is started. A newly started thread has no access to previous declared autoloader.

sandrokeil replied:
Will check

sandrokeil commented:
That's are the [results](https://gist.github.com/sandrokeil/c3e9b94ddadc9e2dfebe462a8f4e0d85). Can anyone confirm that?

sandrokeil commented:
@prolic Optimized Composer autoloader will resolve the class not found issue.

codeliner commented:
@sandrokeil I tried running the benchmark using docker, but I still get:

```
Class 'Prooph\EventStore\Pdo\Exception\RuntimeException' not found
#0 /app/vendor/prooph/pdo-event-store/src/Projection/PdoEventStoreProjector.php(583): Prooph\EventStore\Pdo\Projection\PdoEventStoreProjector->persist()
#1 /app/vendor/prooph/pdo-event-store/src/Projection/PdoEventStoreProjector.php(488): Prooph\EventStore\Pdo\Projection\PdoEventStoreProjector->handleStreamWithSingleHandler('blog-895a15ec-1...', Object(Prooph\EventStore\Pdo\PdoStreamIterator))
```
while running mariadb benchmark

and 

```
Writer 139792754309864 started
Undefined class constant 'DEFAULT_CIPHERS'
#0 /app/vendor/triagens/arangodb/lib/ArangoDBClient/ConnectionOptions.php(205): ArangoDBClient\ConnectionOptions::getDefaults()
#1 /app/vendor/triagens/arangodb/lib/ArangoDBClient/Connection.php(112): ArangoDBClient\ConnectionOptions->__construct(Array)
#2 /app/src/functions.php(86): ArangoDBClient\Connection->__construct(Array)
#3 /app/src/StreamCreator.php(35): Prooph\EventStoreBenchmarks\createConnection('arangodb')
```

while running arrangodb benchmark

sandrokeil commented:
@codeliner The MariaDB issue is #6 I don’t know why this happened. Have you installed the Composer deps with optmized autoloader (`-o` Or `dump-autoloader -o`)? This has fixed the error for me? 🤔

codeliner commented:
yes, I ran composer update with  `-o`

And how did you solve the arangodb error? My real world test breaks immediately for arangodb, so I don't get any numbers from that test

prolic commented:
@codeliner load that class manually before a thread starts.

On Oct 3, 2017 04:26, "Alexander Miertsch" <notifications@github.com> wrote:

> yes, I ran composer update with -o
>
> And how did you solve the arangodb error? My real world test breaks
> immediately for arangodb, so I don't get any numbers from that test
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/prooph/event-store-benchmarks/pull/4#issuecomment-333654429>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAYEvDzzOYEeNboZ65-q9YDihWRs6qB4ks5soUbbgaJpZM4PqbTF>
> .
>

sandrokeil commented:
@prolic / @codeliner I've fixed the class loading issue. I guess the ArangoDB error is special here, because the constant has `null` as value. Loading of classes does not resolve this issue, so I replaced the the constant with `null`.

I'm working on a more accurate benchmark suite by limiting the used system resources, stay tuned.

sandrokeil commented:
@prolic / @codeliner I've limit the service resources for better results. We have to fix the postgres issue #5

sandrokeil commented:
Here are my new results. I would say this can be merged now, and then we can fix the other issues.
[results.txt](https://github.com/prooph/event-store-benchmarks/files/1352411/results.txt)

prolic commented:
Alright, let me review tomorrow.

Sascha-Oliver Prolic

2017-10-03 21:27 GMT+08:00 Sandro Keil <notifications@github.com>:

> Here are my new results. I would say this can be merged now, and then we
> can fix the other issues.
> results.txt
> <https://github.com/prooph/event-store-benchmarks/files/1352411/results.txt>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/prooph/event-store-benchmarks/pull/4#issuecomment-333841298>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAYEvNg-PmI0WhEhyyOs4vsXITwz_rUMks5sojY6gaJpZM4PqbTF>
> .
>

prolic commented:
About data too long for position on mariadb: maybe the json field is too small on mariadb? Can we increase the value or use something else?

prolic commented:
I made some small changes, now we have a merge conflict.

sandrokeil commented:
@prolic I can't get your changes working 
```
Starting benchmark warmup postgres!

Fatal error: Uncaught PDOException: SQLSTATE[08006] [7] FATAL:  database "dev" does not exist in /app/src/functions.php:73
Stack trace:
#0 /app/src/functions.php(73): PDO->__construct('pgsql:host=post...', 'dev', 'dev')
#1 /app/src/functions.php(230): Prooph\EventStoreBenchmarks\createConnection('postgres')
#2 /app/src/benchmark.php(56): Prooph\EventStoreBenchmarks\createConnections()
#3 {main}
  thrown in /app/src/functions.php on line 73
```

prolic commented:
Just rebase, I'll check tomorrow

On Oct 5, 2017 00:40, "Sandro Keil" <notifications@github.com> wrote:

> @prolic <https://github.com/prolic> I can't get your changes working
>
> Starting benchmark warmup postgres!
>
> Fatal error: Uncaught PDOException: SQLSTATE[08006] [7] FATAL:  database "dev" does not exist in /app/src/functions.php:73
> Stack trace:
> #0 /app/src/functions.php(73): PDO->__construct('pgsql:host=post...', 'dev', 'dev')
> #1 /app/src/functions.php(230): Prooph\EventStoreBenchmarks\createConnection('postgres')
> #2 /app/src/benchmark.php(56): Prooph\EventStoreBenchmarks\createConnections()
> #3 {main}
>   thrown in /app/src/functions.php on line 73
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/prooph/event-store-benchmarks/pull/4#issuecomment-334216227>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAYEvMIrqN6RIiFDVKS4gBUK8pOLNOMMks5so7T2gaJpZM4PqbTF>
> .
>
